map-data/src/main/java/org/triplea/map/reader/XmlParser.java

```diff
@@ -0,0 +1,154 @@
+      new GenericParser(currentTag, attributeHandlers, childTagHandlers, body -> {})
+          .parse(streamReader);
+    }
+  }
+
+  @AllArgsConstructor(access = AccessLevel.PRIVATE)
+  public static class BodyParser {
+    private final String currentTag;
+    private final Map<String, Consumer<String>> attributeHandlers;
+    private Consumer<String> bodyHandler;
+
+    public void parse(final XMLStreamReader streamReader) throws XMLStreamException {
+      new GenericParser(currentTag, attributeHandlers, Map.of(), bodyHandler).parse(streamReader);
+    }
+  }
+
+  @Log
+  @AllArgsConstructor(access = AccessLevel.PRIVATE)
+  public static class GenericParser {
+    private final String currentTag;
+    private final Map<String, Consumer<String>> attributeHandlers;
+    private final Map<String, ThrowingRunnable> childTagHandlers;
+    private Consumer<String> bodyHandler;
+
+    public void parse(final XMLStreamReader streamReader) throws XMLStreamException {
+      boolean endTagReached = false;
+
+      log.info("Processing current tag: " + currentTag);
+
+      if (streamReader.getEventType() == XMLStreamReader.START_ELEMENT) {
+        for (int i = 0, n = streamReader.getAttributeCount(); i < n; i++) {
+          final int index = i;
+
+          final String attributeName = streamReader.getAttributeLocalName(index);
+          final Consumer<String> attributeHandler = attributeHandlers.get(attributeName);
+          if (attributeHandler != null) {
+            String attributeValue = streamReader.getAttributeValue(index);
+            attributeHandler.accept(attributeValue);
+          }
+        }
+      }
+
+      while (streamReader.hasNext() && !endTagReached) {
+        log.info("event type: " + streamReader.getEventType());
+        final int event = streamReader.next();
+        switch (event) {
+          case XMLStreamReader.START_ELEMENT:
+            final String childTag = streamReader.getLocalName();
+            log.info("start tag reached: " + childTag);
+            final ThrowingRunnable childTagHandler = childTagHandlers.get(childTag);
+            if (childTagHandler != null) {
+              try {
+                childTagHandler.run();
+              } catch (final Throwable throwable) {
+                throw new XMLStreamException("Parsing failed on tag: " + currentTag, throwable);
+              }
+            }
+            break;
+          case XMLStreamReader.CHARACTERS:
+            if (streamReader.hasText()) {
+              final String text = streamReader.getText();
+              log.info("Characters Event reached: " + text);
+              bodyHandler.accept(text);
+            }
+            break;
+          case XMLStreamReader.END_ELEMENT:
+            final String endTagName = streamReader.getLocalName();
+            log.info("end tag reached: " + endTagName);
+            if (endTagName.equals(currentTag)) {
+              endTagReached = true;
```

Comment:
Nitpick: In this case I'd actually go for a return statement or use labels to break the while loop